design/FeaturesWithLimits.md

```diff
@@ -0,0 +1,124 @@
+  - *Unavailable*:   error, feature not available
+  - *Unimplemented*: error, feature not available
+- `adapter.requestDevice({ requiredFeatures: ["new-feature"],  requiredLimits: { newLimit:                       5 } })`
+  - *Available (7)*: OK, enabled with limit set to 5
+  - *Available (5)*: OK, enabled with limit set to 5
+  - *Unavailable*:   error, feature not available, limit key not available
+  - *Unimplemented*: error, feature not available, limit key not available
+- `adapter.requestDevice({ requiredFeatures: ["new-feature"],  requiredLimits: { newLimit:                       7 } })`
+  - *Available (7)*: OK, enabled with limit set to 7
+  - *Available (5)*: error, limit value not available
+  - *Unavailable*:   error, feature not available, limit key not available
+  - *Unimplemented*: error, feature not available, limit key not available
+- `adapter.requestDevice({ requiredFeatures: adapter.features, requiredLimits: { newLimit:               undefined } })`
+  - *Available (7)*: OK, enabled with limit set to default (5)
+  - *Available (5)*: OK, enabled with limit set to default (5)
+  - *Unavailable*:   OK, neither limit nor feature is requested
+  - *Unimplemented*: OK, neither limit nor feature is requested
+- `adapter.requestDevice({ requiredFeatures: adapter.features, requiredLimits: { newLimit: adapter.limits.newLimit } })`
+  - *Available (7)*: OK, enabled with limit set to 7
+  - *Available (5)*: OK, enabled with limit set to 5
+  - *Unavailable*:   OK, neither limit nor feature is requested
+  - *Unimplemented*: OK, neither limit nor feature is requested
+- `adapter.requestDevice({ requiredFeatures: adapter.features, requiredLimits: { newLimit:                       5 } })`
+  - *Available (7)*: OK, enabled with limit set to 5
+  - *Available (5)*: OK, enabled with limit set to 5
+  - *Unavailable*:   error, limit key not available
+  - *Unimplemented*: error, limit key not available
+- `adapter.requestDevice({ requiredFeatures: adapter.features, requiredLimits: { newLimit:                       7 } })`
+  - *Available (7)*: OK, enabled with limit set to 7
+  - *Available (5)*: error, limit value not available
+  - *Unavailable*:   error, limit key not available
+  - *Unimplemented*: error, limit key not available
+
+TODO: Open questions:
+- Should the device be "reified" such that when the feature is not enabled, the limit will be shown as `undefined`?
+    - We already have to reify such that when the feature is enabled, the limit will be shown as `5`.
+- Or should the device request be "reified" such that when the limit is requested, the feature is implicitly requested?
+    - (I can't figure out if I ever opened an issue about it but I have a proposal bouncing around in my head that we should do some reification of device requests so if you ask for a higher limit, we'll also raise other limits such that it will actually be possible to use it. This would be like that.)
```

Comment:
@toji PTAL, in particular at this - it would be a smallish tweak to the proposal (I haven't worked out all of the case studies for it yet)

Reply:
I'm not in favor of implicitly requesting the feature when an associated limit is requested, for a couple of reasons:

 - There may be multiple limits associated with a feature, and some limits which may potentially be applicable to multiple features. That feels like it could easily become complicated, confusing, and difficult for developers to predict.
 - It would lead to a pattern where someone who simply copies the adapter limits may unintentionally end up turning on features, which doesn't feel desirable. Could end up having an unintended performance hit in some cases, for example, where enabling the feature requires new validation.
 - Having this kind of side effect leads to developers being surprised, which in turn makes for more difficult debugging. ("I know this GPU has a bug with that feature, but I'm definitely not turning it on, see?")